File: OC/内容层动画/WGOCCoreAnimation/WGOCCoreAnimation/渐进圆环效果/WGButtnAniView.m
//
//  WGButtnAniView.m
//  WGOCCoreAnimation
//
//  Created by wanggang on 2019/5/22.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGButtnAniView.h"
#import "WGCircleAniView.h"

@interface WGButtnAniView()<CAAnimationDelegate, WGCircleAniViewDelegate>

@property (nonatomic, strong) UIView *contentView;
@property (nonatomic, strong) WGCircleAniView *circleView;

@end

@implementation WGButtnAniView

@synthesize firstColor = _firstColor;

-(instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        
        [self buildBaseView];
    }
    return self;
}

- (void)buildBaseView{
    
    self.contentView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.width, self.height)];
    _contentView.backgroundColor = self.firstColor;
    [self addSubview:_contentView];
    
    self.circleView = [[WGCircleAniView alloc] initWithFrame:CGRectMake(0, 0, self.width, self.height)];
    _circleView.circleAniViewDelegate = self;
    [self addSubview:_circleView];
    
    self.titleLab = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, self.width, self.height)];
    _titleLab.font = [UIFont systemFontOfSize:20];
    _titleLab.textColor = [UIColor blackColor];
    _titleLab.textAlignment = NSTextAlignmentCenter;
    [self bringSubviewToFront:_titleLab];
    [self addSubview:_titleLab];
}

//返回基础动画
- (CABasicAnimation *)aniKeyPath:(NSString *)keyPath fromValue:(id)fromVal toValue:(id)toVal{
    
    CABasicAnimation *ani = [[CABasicAnimation alloc] init];
    ani.keyPath = keyPath;
    ani.fromValue = fromVal;
    ani.toValue = toVal;
    ani.fillMode = kCAFillModeForwards;
    [ani setRemovedOnCompletion:NO];
    return ani;
}

//变圆动画
- (void)startAniDuation{
    
    self.titleLab.hidden = YES;
    self.userInteractionEnabled = NO;
    
    //调整大小
    NSValue *to = [NSValue valueWithCGRect:CGRectMake(self.x+(self.width-self.height)/2, self.height, self.height, self.height)];
    CABasicAnimation *boundsAni = [self aniKeyPath:@"bounds" fromValue:nil toValue:to];
    
    //背景色渐变透明
    CABasicAnimation *backgroundColorAni = [self aniKeyPath:@"backgroundColor" fromValue:(id)self.firstColor.CGColor toValue:(id)[UIColor clearColor].CGColor];
    
    //变圆
    CABasicAnimation *circleAni = [self aniKeyPath:@"cornerRadius" fromValue:@(0) toValue:@(self.height/2)];
    
    //边框宽度
    CABasicAnimation *borderWidthAni = [self aniKeyPath:@"borderWidth" fromValue:@(0) toValue:@(5)];
    
    //边框颜色
    CABasicAnimation *borderColorAni = [self aniKeyPath:@"borderColor" fromValue:(id)self.firstColor.CGColor toValue:(id)[UIColor lightGrayColor].CGColor];
    
    //动画组
    CAAnimationGroup *contentGroup = [[CAAnimationGroup alloc] init];
    contentGroup.animations = @[boundsAni, backgroundColorAni, circleAni, borderColorAni, borderWidthAni];
    contentGroup.duration = self.firstDuation;
    contentGroup.fillMode = kCAFillModeForwards;
    [contentGroup setRemovedOnCompletion:NO];
    contentGroup.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut];
    contentGroup.delegate = self;
    [contentGroup setValue:@"startAnimationValue" forKey:@"startAnimationKey"];
    
    [self.contentView.layer addAnimation:contentGroup forKey:@"startAnimationValue"];
}

//回复原状动画
- (void)resetAniDuation{
    
    //调整大小
    NSValue *to = [NSValue valueWithCGRect:CGRectMake(0, 0, self.width, self.height)];
    CABasicAnimation *boundsAni = [self aniKeyPath:@"bounds" fromValue:nil toValue:to];
    
    //恢复背景色
    CABasicAnimation *backrroundColorAni = [self aniKeyPath:@"backgroundColor" fromValue:nil toValue:(id)self.thirdColor.CGColor];
    
    //去圆角
    CABasicAnimation *circleAni = [self aniKeyPath:@"cornerRadius" fromValue:nil toValue:@(0)];
    
    //边框宽度为0
    CABasicAnimation *borderWidthAni = [self aniKeyPath:@"borderWidth" fromValue:nil toValue:@(0)];
    
    CAAnimationGroup *group = [[CAAnimationGroup alloc] init];
    group.animations = @[boundsAni, backrroundColorAni, circleAni, borderWidthAni];
    group.duration = self.thirdDuation;
    group.fillMode = kCAFillModeForwards;
    [group setRemovedOnCompletion:NO];
    group.delegate = self;
    [group setValue:@"resetAnimationValue" forKey:@"resetAnimationKey"];
    [self.contentView.layer addAnimation:group forKey:@"resetAnimationValue"];
    
}

#pragma mark --<CAAnimationDelegate>方法
-(void)animationDidStop:(CAAnimation *)anim finished:(BOOL)flag{
    
    if (flag) {
        
        NSString *animType = [anim valueForKey:@"startAnimationKey"];
        if (animType && [animType isEqualToString:@"startAnimationValue"]) {
            //开始进度动画
            [self.circleView startCircleAniColor:self.secondColor duation:self.secondDuation];
        }else{
            NSString *animType = [anim valueForKey:@"resetAnimationKey"];
            if (animType && [animType isEqualToString:@"resetAnimationValue"]) {
                //回复动画完成
                self.userInteractionEnabled = YES;
                self.titleLab.hidden = NO;
                self.titleLab.text = @"完成";
            }
        }
    }
}

#pragma mark --<WGCircleAniViewDelegate>方法
-(void)aniDidStop:(CAAnimation *)ani finished:(BOOL)finished{
    
    if (finished) {
        //恢复动画
        [self resetAniDuation];
    }
}

//控制背景色和contentView一致
-(void)setBackgroundColor:(UIColor *)backgroundColor{
    
    self.firstColor = backgroundColor;
    self.contentView.backgroundColor = self.firstColor;
}

- (UIColor *)firstColor{
    if (!_firstColor) {
        _firstColor = UIColor.cyanColor;
    }
    return _firstColor;
}

-(void)setFirstColor:(UIColor *)firstColor{
    
    _firstColor = firstColor;
    self.contentView.backgroundColor = _firstColor;
}

- (CGFloat)firstDuation{
    if (_firstDuation <= 0) {
        _firstDuation = 3;
    }
    return _firstDuation;
}

- (UIColor *)secondColor{
    if (!_secondColor) {
        _secondColor = UIColor.redColor;
    }
    return _secondColor;
}

- (CGFloat)secondDuation{
    if (_secondDuation <= 0) {
        _secondDuation = 3;
    }
    return _secondDuation;
}

- (UIColor *)thirdColor{
    if (!_thirdColor) {
        _thirdColor = UIColor.cyanColor;
    }
    return _thirdColor;
}

- (CGFloat)thirdDuation{
    if (_thirdDuation <= 0) {
        _thirdDuation = 3;
    }
    return _thirdDuation;
}

@end

File: Swift/内容层动画/WGCoreAnimation/WGCoreAnimation/渐进圆环效果/WGButtnAniView.swift
//
//  WGButtnAniView.swift
//  WGCoreAnimation
//
//  Created by wanggang on 2019/4/24.
//  Copyright © 2019 wanggang. All rights reserved.
//

import UIKit

class WGButtnAniView: UIView {

    var startColor = UIColor.cyan
    var finishColor = UIColor.yellow
    
    //内容视图
    var contentView: UIView?
    //边框视图
    var borderView: UIView?
    //标题
    var titleLabel: UILabel?
    //展示进度的视图
    var circleView: WGCircleAniView?

    var btnX: CGFloat = 0;
    var btnY: CGFloat = 0;
    var btnW: CGFloat = 0;
    var btnH: CGFloat = 0;
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        btnX = self.frame.origin.x
        btnY = self.frame.origin.y
        btnW = self.frame.size.width
        btnH = self.frame.size.height
        
        contentView = UIView(frame: CGRect(x: 0, y: 0, width: btnW, height: btnH))
        contentView?.backgroundColor = startColor
        self.addSubview(contentView!)
        
        borderView = UIView(frame: CGRect(x: 0, y: 0, width: btnW, height: btnH))
        borderView?.backgroundColor = UIColor.clear
        borderView?.layer.borderWidth = 3
        borderView?.layer.borderColor = startColor.cgColor
        self.addSubview(borderView!)
        
        titleLabel = UILabel(frame: CGRect(x: 0, y: 0, width: btnW, height: btnH))
        titleLabel?.text = "开始"
        titleLabel?.textColor = UIColor.black
        titleLabel?.backgroundColor = UIColor.clear
        titleLabel?.textAlignment = .center
        titleLabel?.font = UIFont.systemFont(ofSize: 20)
        self.addSubview(titleLabel!)
        
        circleView = WGCircleAniView(frame: CGRect(x: 0, y: 0, width: btnW, height: btnH))
        circleView?.wgCircleAniViewDelegate = self
        self.addSubview(circleView!)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    //返回动画
    func getBasicAnimation(keyPath: String, fromValue: Any?, toValue: Any?) -> CABasicAnimation {
        
        let ani = CABasicAnimation()
        ani.keyPath = keyPath
        ani.fromValue = fromValue
        ani.toValue = toValue
        return ani
    }
    
    func startAnimation() {
        
        /*
         由 按钮->透明的灰色框  动画动画可拆解为两部分
         一:contentView渐变成一个圆且淡出
         二:borderView渐变成一个圆框且变成灰色
         */
        self.isUserInteractionEnabled = false
        titleLabel?.isHidden = true
        //调整大小
        let toValue = NSValue(cgRect: CGRect(x: btnX+(btnW-btnH)/2, y: btnH, width: btnH, height: btnH))
        let boundsAni = getBasicAnimation(keyPath: "bounds", fromValue: nil, toValue: toValue)
        
        //圆角
        let cornerRadiusAni = getBasicAnimation(keyPath: "cornerRadius", fromValue: 0, toValue: btnH/2)
        
        //淡出
        let opacityAni = getBasicAnimation(keyPath: "opacity", fromValue: nil, toValue: 0)
        
        //动画组:contentView渐变成一个圆且淡出
        let conAniGroup = CAAnimationGroup()
        conAniGroup.animations = [boundsAni, cornerRadiusAni, opacityAni]
        conAniGroup.duration = 1
        conAniGroup.repeatCount = 1
        conAniGroup.fillMode = .forwards
        conAniGroup.isRemovedOnCompletion = false
        conAniGroup.timingFunction = CAMediaTimingFunction(name: .easeInEaseOut)
        contentView?.layer.add(conAniGroup, forKey: nil)
        
        //边框颜色渐变
        let borderColorAni = getBasicAnimation(keyPath: "borderColor", fromValue: nil, toValue: UIColor.lightGray.cgColor)
        
        //动画组:borderView渐变成一个圆框且变成灰色
        let borAniGroup = CAAnimationGroup()
        borAniGroup.animations = [boundsAni, cornerRadiusAni, borderColorAni]
        borAniGroup.duration = 1
        borAniGroup.repeatCount = 1
        borAniGroup.fillMode = .forwards
        borAniGroup.isRemovedOnCompletion = false
        borAniGroup.timingFunction = CAMediaTimingFunction(name: .easeInEaseOut)
        borAniGroup.delegate = self
        borAniGroup.setValue("startAnimationValue", forKey: "startAnimationKey")
        borderView?.layer.add(borAniGroup, forKey: "startAnimationValue")
        
    }
    
    func startAnimationReset() {
        
        self.isUserInteractionEnabled = false
        //调整大小
        let fromValue = NSValue(cgRect: CGRect(x: btnX+(btnW-btnH)/2, y: btnH, width: btnH, height: btnH))
        let toValue = NSValue(cgRect: CGRect(x: 0, y: 0, width: btnW, height: btnH))
        let boundsAni = getBasicAnimation(keyPath: "bounds", fromValue: fromValue, toValue: toValue)
        
        //去圆角
        let cornerRadiusAni = getBasicAnimation(keyPath: "cornerRadius", fromValue: btnH/2, toValue: 0)
        
        //淡入
        let opacityAni = getBasicAnimation(keyPath: "opacity", fromValue: nil, toValue: 1)
        
        //改变下背景色,区别已完成
        let backGroundColorAni = getBasicAnimation(keyPath: "backgroundColor", fromValue: nil, toValue: finishColor.cgColor)
        
        //动画组:contentView渐变成原来形状且改变背景色
        let conAniGroup = CAAnimationGroup()
        conAniGroup.animations = [boundsAni, cornerRadiusAni, opacityAni, backGroundColorAni]
        conAniGroup.duration = 3
        conAniGroup.repeatCount = 1
        conAniGroup.fillMode = .forwards
        conAniGroup.isRemovedOnCompletion = false
        conAniGroup.timingFunction = CAMediaTimingFunction(name: .easeInEaseOut)
        contentView?.layer.add(conAniGroup, forKey: nil)
        
        //边框颜色渐变
        let borderColorAni = getBasicAnimation(keyPath: "borderColor", fromValue: nil, toValue: finishColor.cgColor)
        
        //动画组:borderView由灰色框恢复为原形状
        let borAniGroup = CAAnimationGroup()
        borAniGroup.animations = [boundsAni, cornerRadiusAni, borderColorAni]
        borAniGroup.duration = 3
        borAniGroup.repeatCount = 1
        borAniGroup.fillMode = .forwards
        borAniGroup.isRemovedOnCompletion = false
        borAniGroup.timingFunction = CAMediaTimingFunction(name: .easeInEaseOut)
        borAniGroup.delegate = self
        borAniGroup.setValue("startAnimationResetValue", forKey: "startAnimationResetKey")
        borderView?.layer.add(borAniGroup, forKey: "startAnimationResetValue")
        
    }
    
}

//动画代理
extension WGButtnAniView: CAAnimationDelegate{
    
    func animationDidStop(_ anim: CAAnimation, finished flag: Bool) {
        
        if flag {
            let aniType = anim.value(forKey: "startAnimationKey")
            let aniType01 = anim.value(forKey: "startAnimationResetKey")
            if let type = aniType {
                if (type as! NSString).isEqual(to: "startAnimationValue"){
                    //开始进度动画
                    circleView?.startCircleAnimation()
                }
            }else if let type = aniType01 {
                
                if (type as! NSString).isEqual(to: "startAnimationResetValue"){
                    
                    self.isUserInteractionEnabled = true
                    titleLabel?.text = "完成"
                    titleLabel?.isHidden = false
                }
            }
        }
    }
}

//进度圆环代理
extension WGButtnAniView: WGCircleAniViewDelegate{
    
    func circleAniStop() {
        //还原之前形状
        circleView?.removeFromSuperview()
        startAnimationReset()
    }
}
